docs/stdout-exporter-design.md

```diff
@@ -0,0 +1,223 @@
+# StdoutExporter Design
+
+In strongly typed languages typically there will be 2 separate `Exporter` interfaces, one that accepts spans from a tracer (SpanExporter) and one that accepts metrics (MetricsExporter)
+
+The exporter SHOULD be called with a checkpoint of finished (possibly dimensionally reduced) export records. Most configuration decisions have been made before the exporter is invoked, including which instruments are enabled, which concrete aggregator types to use, and which dimensions to aggregate by.
+
+## Use Cases
+
+Monitoring and alerting systems commonly use the data provided through metric events or tracers, after applying various aggregations and converting into various exposition format. After getting the data, the systems need to be able to see the data. The StdoutExporter will be used here to print data through Stdout, this is seen as a simple exporter where the user doesn’t have the burden of implementing or setting up a protocol dependent exporter.
+
+The StdoutExporter will also be used as a debugging tool for the Metrics API/SDK and Tracing API/SDK which are currently work in progress projects. This exporter will allow contributors to easily diagnose problems when working on the project.
+
+## Push vs Pull Exporter
+
+There are two different versions of exporters: Push and Pull. A Push Exporter pushes the data outwards towards a system, in the case of the StdoutExporter it prints data into Stdout. A Pull Exporter exposes data to some endpoint for another system to grab the data. 
+
+The StdoutExporter will only be implementing a Push Exporter framework.
+
+## Design Tenets
+
+* Reliability
+    * The Exporter should be reliable; data exported should always be accounted for. The data will either all be successfully exported to the destination server, or in the case of failure, the data is dropped. `Export` will always return failure or success to notify the user of the result.
+    * Thread Safety
+        * Export() should never be called concurrently. This is handled by the controller in the StdoutMetricsExporter.
```

Comment:
I don't know if this is true. Span Exporter could be called concurrently when as span ends from multiple threads.

Reply:
I took this directly from the spec, see - https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#span-exporter

Reply:
This should be true. A span exporter is only called by a single processor, and the processor should make sure that there are no concurrent calls to the `Export` method.

With a batched span processor, `OnEnd` can be called concurrently from different threads, but the processor itself calls `Export` on the processor sequentially.

I see that our simple span processor doesn't conform to that and might end up calling `Export` from different threads. I'll probably open up a PR with a fix for that.

Reply:
Interesting, this is very different from what the SDK has implemented.

https://github.com/open-telemetry/opentelemetry-python/blob/ca8c0970f88b083b57c85322c997595d0ffc3702/docs/examples/basic_tracer/basic_trace.py

I will need to clarity this with the specification SIG.

Reply:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#simple-processor

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#span-processor

If we have a `Simple Span Processor` sending data to a `Console Exporter`, the console exporter would called since the spec mentioned "passes finished spans and passes the export-friendly span data representation to the configured SpanExporter, as soon as they are finished."